### PR TITLE
Bumps Node requirements from 14 to 18

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

We're still supporting Node 14, but it has reached end of life. Node 16 is still maintained, but will reach an early end of life in October, so I think it's reasonable to drop it now rather than publish a major release just for that.

**How did you fix it?**

Bumps the requirements from `14.16` to `18.12` (first LTS from the 18.x release line).

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Review by merceyz — CHANGES_REQUESTED** [review]
I don't see the need of dropping support for these Node.js versions as it doesn't cost us much to keep support for them.
If that changes we can drop them and release a new major version at any point.

> rather than publish a major release just for that

It's just a number.

**Comment by arcanis** [comment]
The thing is, we didn't treat it as just a number so far, and I think it's more than that in the mind of (at least a part of) the community. There's an argument to be made to change that, but it has some ramifications and I'd prefer not to rush it and repeat the kind of misunderstanding from the v2 release (in other words, that seems a post-4.0 discussion to me).

As for upgrading Node, I'm fine with keeping a loose requirement on the CLI check, but I'd prefer to avoid keeping an outdated transpilation for the sake of Node 14 users. Node 16 is perhaps another story since it has a few more months, but considering it'll quickly become unsafe (due to the OpenSSL release line [falling out of support](https://nodejs.org/en/blog/announcements/nodejs16-eol), causing Node to do the same), it could even be seen as encouraging a good practice.

**Comment by arcanis** [comment]
I removed the reverts on logic updates, but kept the test runtime upgrade & `engines` field upgrade.

**Comment by merceyz** [comment]
If you absolutely want to drop support for Node.js <18 and push this through we should do it fully, not half-way as that's confusing.

**Comment by nicolo-ribaudo** [comment]
> For example Babel, depending on the outcome of https://github.com/babel/babel/pull/15585.

Note that we first run Yarn and then change the Node.js version, so this wouldn't affect us :) We currently use Yarn 3 even if Babel still supports Node.js 6.

**Comment by arcanis** [comment]
> If you absolutely want to drop support for Node.js <18 we should do it fully, not half-way as that's confusing.

Let's do it then; since Yarn is versioned by project, it shouldn't break setups outside of upgrade attempts, which are anyone subject to other breaking changes. Worst case, if someone has a very valid use case, we'll be able to revert it without having to release a new major. I'll add back the original changes.

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ Yarn now accepts sponsorships! Please give a look at our [OpenCollective](https:
 
 ### **Major Changes**
 
+- With Node.js 16's End of Life [approaching fast](https://nodejs.org/en/blog/announcements/nodejs16-eol), we dropped support for both Node.js 14 and 16, with Node.js 18 being the recommended LTS version.
```
Comment (merceyz) [inline CHANGELOG.md:13]:
```suggestion
- With Node.js 16's End of Life [approaching fast](https://nodejs.org/en/blog/announcements/nodejs16-eol), we dropped support for Node.js 12, 13, 14, 15, 16, and 17, with Node.js 18 being the recommended LTS version.
```
There isn't a changelog entry for it but we also dropped support for <14.15 in https://github.com/yarnpkg/berry/pull/4196.

**Review by merceyz — CHANGES_REQUESTED** [review]
A few more files needs to be updated, mainly the `rollup.config.js` files, but see the list from https://github.com/yarnpkg/berry/pull/4196#pullrequestreview-902783965.

File: packages/yarnpkg-pnp/sources/esm-loader/loaderFlags.ts [hunk 2]
```diff
@@ -1,13 +1,10 @@
...
-// The hooks were consolidated in https://github.com/nodejs/node/pull/37468
-export const HAS_CONSOLIDATED_HOOKS = major > 16 || (major === 16 && minor >= 12);
-
 // JSON modules were unflagged in https://github.com/nodejs/node/pull/41736
-export const HAS_UNFLAGGED_JSON_MODULES = major > 17 || (major === 17 && minor >= 5) || (major === 16 && minor >= 15);
+export const HAS_UNFLAGGED_JSON_MODULES = major > 17 || (major === 17 && minor >= 5);
 
 // JSON modules requires import assertions after https://github.com/nodejs/node/pull/40250
-export const HAS_JSON_IMPORT_ASSERTION_REQUIREMENT = major > 17 || (major === 17 && minor >= 1) || (major === 16 && minor > 14);
+export const HAS_JSON_IMPORT_ASSERTION_REQUIREMENT = major > 17 || (major === 17 && minor >= 1);
```
Comment (merceyz) [inline packages/yarnpkg-pnp/sources/esm-loader/loaderFlags.ts:7]: Should be able to remove these flags since they'll always be true.

**Review by merceyz — CHANGES_REQUESTED** [review]

**Review by merceyz — APPROVED** [review]
LGTM, though there is a merge conflict that needs to be resolved.

**Comment by merceyz** [comment]
> Internal Error: spawn ETXTBSY
https://github.com/yarnpkg/berry/actions/runs/5350638693/jobs/9703540026?pr=5445#step:5:1067

Should be fixed in https://github.com/libuv/libuv/pull/4059 but that isn't released yet so we can either disable io_uring support by setting `UV_USE_IO_URING=0` in the environment or lock the CI to Node.js v20.2.0 until the fix is out.
I vote for the latter.

File: .github/workflows/integration-workflow.yml [hunk 3]
```diff
@@ -201,19 +201,19 @@ jobs:
       fail-fast: false
       matrix:
         # We run the ubuntu tests on multiple Node versions with 2 shards since they're the fastest.
-        node: [14, 16, 18, 19]
+        node: [18, 19, 20.2]
```
Comment (paul-soporan) [inline .github/workflows/integration-workflow.yml:204]: We should leave a comment about it.

**Comment by arcanis** [comment]
Branch protection rules updated to the new job names

**Pull request merged**